### DISCUSSION TITLE
Simplify settings management in multi-player environments

### DISCUSSION THREAD

**Discussion by madbrain76** [music-assistant/support#2501] [Feature requests and ideas]
I'll start by prefacing that if you only have a single player device, the current user experience for player is perfectly fine.
However, when you have 24 physical devices, and 12 of them support 3 protocols, they show up as 48 different players in MA.
While I can of course disable those with duplicate protocols, the player count will never be below 24. Also, player counts can change over time when software renderers are installed.

The issue is that the default player settings are not well suited for classical music or my tastes. There are 3 main problems :

1. gapless support is not enabled by default for all devices that support it.
While gapless albums are a small minority of albums, some of my favorite albums require gapless playback.
Of the 3 player providers I have tried - Airplay, Chromecast and DLNA, only Airplay does gapless by default. That is because Apple designed it properly to support gapless from the ground up. Chromecast and DLNA didn't, or left the gapless feature optional.  It is amazing that Music Assistant managed to provide gapless support on these 2 providers as well. It isn't enabled by default, though. It requires enabling "queue flow" mode on each player. Doing on 12-36 individual players one at a time is not ideal.

2. EBU-R128 volume normalization is enabled by default.
This specification was originally designed for broadcast radio and TV, to help with the problem of commercials getting too loud.
It wasn't meant to be used for mastering music. Regardless of its original goal, using EBU-R128 can have very undesirable effects, if applied separately at the track level on an album, or on separate discs that form a single album. For example, I listen to a lot of solo harpsichord music. I also have one at home, play it, and record it. One needs to be very careful about mastering. I wasn't part of this discussion, but it will shed some light on it - the gist is that mastering engineers spend a lot of time mastering harpsichord albums, and EBU-R128 unfortunately interferes with that.
https://gearspace.com/board/remote-possibilities-in-recording-amp-production/993904-harpsichord-mastering-monitoring-loudness-levels.html
The harpsichord is an instrument with limited volume control - the only way to really vary the volume is to play chords, or enable more stops, if available (mine has 3 stops). In general, the volume range is fairly narrow. However, the peak can vary quite a bit from one track to another. And that's when the volume normalization really hurts. It can result in 2 tracks having significantly different volume, for the same solo instrument. When this happens on 2 tracks of a gapless album, in the middle of a note, it is a very unnatural experience.
Manually disabling this on 48 different players is quite a task.

3. fade-in and crossfade effects
These effects are also very unnatural on a harpsichord, since its level fundamentally cannot change. 
While the "enable crossfade" feature is thankfully off by default, there are some situations in MA where fade-in effects are added. At least one case of this is currently being worked on to make it optional. I would like to be able to disable fade-in globally for all cases, and for all providers and players.

I don't really have a clear suggestion on how the UI should look like. The minimalist approach would be to have a single global checkbox to "Preserve musical integrity mode", or any other more self-explanatory wording, grey out all the relevant settings in all players, existing and future, in all music providers, and call it a day. It could be called "Classical/Baroque music mode", or even "Harpsichord mode", though I doubt the later will fly :)

The reality is that these issues can affect many types of albums. While I wish the default settings were different so I didn't have to change them so many times. I realize others have different preferences, so I'm not asking for defaults to be changed, only for a simpler way to change them on multiple players.

**Comment by marcelveldt** [comment]
We are focused on mainstream, so with feature requests we follow what most people seem to prefer/like.
At the same time we are only a small team of volunteers working on this so we cant implement every little detail, we focus on the bigger picture and if we, at one day, get bored, we can spend time on the small details. Not that I don't appreciate your input but I'd like to realistically warn you that it may take a lot of time before any of it gets looked at, unless a lot of other classical musical listeners step in and start voting here ;-)

Maybe you should have a look at Roon (www.roonlabs.com) as that is a better fit for classic musical listeners, considering the type of feature requests you have. It will probably take years before we even come close to the level of detail you are seeking.

1) We do not enable flow mode by default because it comes at a cost of on-device metadata on many devices. 
Feature-wise I agree we should probably enable it by default but there are DLNA players that are perfectly capable of doing gapless out of the box. Problem is that the DLNA spec/implementation is such a mess that its just guesswork.

2) With the loudness war where streaming providers master their music at insane levels and that combined with the fact that MA can have multiple sources, we NEED volume normalization by default. 
But we implemented volume normalization on the proper way, so not by compressing dynamic range. It should work with all music types. It is done using the loudnorm filter within ffmpeg, implemented by a very talented sound engineer. I think you should first investigate that (or listen to it!!) before making bold claims here. 

3) No need to replicate this report in 100 different places. We get your point by now and it will get looked at some point. So far you are literally the only one noticing this so I guess its not that big of a problem. Remember that this only happens at resume (from stop) of music . That its happening for you also at unpause is a bug. 

On another note:
If you have 48 players you either live in some castle or you should consider getting a different hobby ;-)
In other words: I think most people have just a few players and managing those few per-player setting is worth the effort. It should be a one-time thing. Also you should probably choose one of the streaming protocols and use that. Why have both the dlna and airplay variants of each player and not just choose to stick with Airplay alone.

FYI: I'm thinking about introducing some way to "combine" multiple players into one. For example for a Sonos player that also accepts Airplay. For now this is just an idea still forming in my head and not yet implemented.

**Comment by madbrain76** [comment]
@marcelveldt ,

I can completely appreciate that the team is small and things have to be prioritized. I'm not expecting this to be dealt with in the near future.
I'm running into this a lot because I have been, as you know, testing the limits of MA. And that involves removing and re-adding the player providers, music providers, add-on, etc - quite a lot. Once in a while I try to actually listen to music when not testing ;)

I did in fact look at Roon. It did a good job with cloud, but with my local library, it tried to use its own metadata, which wasn't always correct. And it doesn't in itself provide any way to manually edit the metadata. The whole thing was just too opaque. It's similar to the way Plex works - it uses some of the existing metadata, and then tries to combine it with its own. The results are sometimes good, but sometimes not so good also. I believe I have the proper tool to edit library metadata already, and having something that just uses it without trying to override it, as MA does, suits me better. I didn't love the price of Roon either, to be fair, but that wouldn't have stopped me if I really liked it.
Having worked on open-source a fair bit of my career, an open source project really appeals to me, also.

1. I see. Yes, some of my devices handle gapless just fine without the queue flow mode, but others don't.
2. I did listen to it last month, and it produced very objectionable results. I don't remember which album it was on, that info may be in the post. And there is one album in particular I use to test gapless. . I tried to reproduce the issue with the current MA and this disc last week-end, and this time, it didn't occur. Did something change recently ?
At the risk of posting yet another feature request, maybe this could be done conditionally based on the provider - if some are mastering more aggressively than others - but most likely it's just album-specific.
3. sorry to mention it again

It's not a castle, but it is a really big house. It came with 32 built-in speakers in the ceilings, walls, etc. And we have have many others. The home theater has as an 11.4 system, bedroom 9.2, office 4.1, guest bedroom 5.1 . And did I mention I have 246 devices Home Assistant, and over 1300 entities ?

And I agree with you about not enabling all the duplicate players. I have 5 devices that are Chromecast-only (2 audio, 3 video). 3 Marantz devices that are Airplay or DLNA. And 12 WiiM devices that are Airplay/DLNA/Chromecast. And there are a few software renderers also that are DLNA (JRiver, Kodi). Unfortunately, there is no single protocol in common, so I must have at least 2 providers enabled, or some devices will be missing. The most reliable protocol for gapless is Airplay. But the 5 Google devices don't support it. Neither does JRiver. There is Airplay support in Kodi, but I haven't tested it yet. With multiple protocols enabled, the duplicates from WiiM will be showing. I proposed a change to separate the enabled from disabled players in the list.

https://github.com/orgs/music-assistant/discussions/2511

**Reply by OzGav** [reply]
On point 2 MA takes readings of the tracks as it plays them for normalisation. So after that first play you should expect a good outcome.

**Comment by madbrain76** [comment]
Ah, I see. That might very well explain why I couldn't reproduce it. Is there a way to reset that to test it again ? Remove the music provider ? Or remove the add-on ?

**Reply by OzGav** [reply]
I would assume that info is stored in the database so probably have to remove the whole addon

**Reply by marcelveldt** [reply]
First play is indeed "dynamic mode" it reads the file dynamics while playing it back and apply a stage 1 normalization - outcome of that is often a bit meh, it helps but I can imagine it will have issues with high dynamic range music. Luckily once it has played a file once, it stores the measurement persistently and all subsequent plays will use the measurement from stage 1.

Maybe what we could do at some point is offer the option to pre-read local files for the measurement

**Comment by madbrain76** [comment]
@OzGav ,
Thanks, I'm trying to reproduce now after deleting the add-on, but still failing. It may not be the right album to test.

@marcelveldt ,
One thing I could suggest is to use metadata already in the file, if any, to determine the dynamics. You might already be doing this with the Musicbrainz metadata. But It seems that there are 5 or 6 different metadata standards for storing this information, unfortunately. I'm perfectly willing to run the audio analysis on my entire local collection in JRiver, which I use to manage my collection. It does have this feature. It will take a bit of time, but I have powerful enough computers and that is perfectly OK. Doing on the MA side on the entire collection may not be ideal, as HA is often running on low-power single board computers. This is not my case, but I still prefer to have the metadata written to the individual files rather than in the MA DB, in case there is ever any issue with that DB..

**Comment by madbrain76** [comment]
@OzGav , spent quite a bit of time today trying to reproduce the issue - no dice. I wonder if it the behavior varies depending on the player provider and its configuration (maybe the queue_flow). And possibly the volume is different between music providers as well.

I finally found the original post that you helped me with . It was in the HA community. The album is listed, and it says I was playing the album from Qobuz. The music provider isn't listed, but it says it happened in gapless mode. The post mentions Chromecast and DLNA, but not Airplay. Hopefully, that will be enough to reproduce. 
https://community.home-assistant.io/t/music-assistant-2-0-your-music-your-players/727626/355

**Comment by madbrain76** [comment]
@marcelveldt @OzGav 
I just reproduced it. It's the opposite of my other report - it went from quiet to loud this time.

Here is exactly what I did. There may be fewer steps required to reproduce, but these are what led to it :

1. add Chromecast player provider
2. set queue flow on the player
3. favorite Byron Schenkman's Harpsichord Variations in the Qobuz app
4. add Qobuz provider
5. select the player from step 2
6. select the album
7. starts playing the album (blue play button)
8. seek near the end of the track
9. let it play through the second track
10. there is no volume issue
11. click on track 1 of the album
12. seek near the end
13. it skips track 2 and goes to track 3 (not clear why - but separate issue)
14. go to the player
15. clear the queue
16. go back to the album view
17. play the album with the blue button
18. seek near the end of the track and let it play through
19. track 2 plays a lot louder than track 1
20. I am able to reproduce by repeating steps 14-19
21. I'm letting it play through, and there seems to be volume differences between many tracks. They don't have any pause, everything is continuous, so it's quite noticeable.

I do have a log, but there doesn't seem to be much of any relevance in it. Maybe it needs a higher level of logging.

Edit: after reinstalling the add-on, I reproduced the above steps exactly with the file system (remote) provider and my local copy of this album. So, the issue isn't music provider specific.

I also let it play through - several times actually.

I then restarted playback of the album on a different player, also a Chromecast with queue flow on.
This time, volume went down very slightly between tracks 1 and 2 - it might not be obvious if you aren't familiar with the harpsichord or the album

However, volume then went up significantly and unmistakably between tracks 2 and 3.

I didn't detect a volume change between tracks 3 and 4 because there is an actual gap, but there could be a small one.

Massive difference between tracks 4 and 5. At the end of track 4, you can hear the mic noise that should be so low-level that it is never heard. Then, track 5 is much quieter.

It seems like the normalization is happening at the track level rather than album level. And that just doesn't work for a gapless harpsichord album, which happens to be one of my favorites. This is something I ran into the very first time I installed MA, and tried to play this album.

The only fix I am aware of in order to be able to listen to this album properly is to disable the volume normalization on the player. If there is another one, I would love to know.

I do listen to a ton of harpsichord, but most of the albums are not gapless, and the volume issue may not be quite as obvious as it is on this album. It also depends how they were mastered.
